common/metadata_info.go

```diff
@@ -0,0 +1,221 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package common
+
+import (
+	"fmt"
+	"github.com/apache/dubbo-go/common/constant"
+	gxset "github.com/dubbogo/gost/container/set"
+	"go.uber.org/atomic"
+	"hash/crc32"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+var IncludeKeys = gxset.NewSet(
+	constant.APPLICATION_KEY,
+	constant.GROUP_KEY, constant.TIMESTAMP_KEY, constant.SERIALIZATION_KEY, constant.CLUSTER_KEY,
+	constant.LOADBALANCE_KEY, constant.PATH_KEY, constant.TIMEOUT_KEY,
+	constant.TOKEN_KEY, constant.VERSION_KEY, constant.WARMUP_KEY,
+	constant.WEIGHT_KEY, constant.RELEASE_KEY)
+
+type MetadataInfo struct {
+	App      string                  `json:"app"`
+	Revision string                  `json:"revision"`
+	Services map[string]*ServiceInfo `json:"services"`
+
+	reported *atomic.Bool `json:"-"`
+}
+
+func NewMetadataInfWithApp(app string) *MetadataInfo {
+	return NewMetadataInfo(app, "", make(map[string]*ServiceInfo))
+}
+
+func NewMetadataInfo(app string, revision string, services map[string]*ServiceInfo) *MetadataInfo {
+	return &MetadataInfo{
+		App:      app,
+		Revision: revision,
+		Services: services,
+		reported: atomic.NewBool(false),
```

Comment:
as I know the default value of a  uber atomic bool is false https://github.com/uber-go/atomic/blob/9c79392530c3be3273bd71db21239f2a5146c10c/bool.go#L41. so u need not to set its value again. pls delete this line.